Add docs on pre/post processing operations and dlq support (#26772)
apache/beam: website/www/site/content/en/documentation/sdks/python-machine-learning.md

For more information on resource hints, see [Resource hints](/documentation/runtime/resource-hints/).

## RunInference Patterns

> This section suggests patterns and best practices that can be used to make your inference pipelines simpler,

Reviewer: ```suggestion
This section suggests patterns and best practices that you can use to make your inference pipelines simpler,
```

Author: Thanks!

For more information on resource hints, see [Resource hints](/documentation/runtime/resource-hints/).

> ## RunInference Patterns

Reviewer: We should include text after this header describing the section. Something along the lines of:

This section provides examples of RunInference patterns.

(But hopefully more informative than that.)

Author: Done - let me know if you think it sounds ok: "This section suggests patterns and best practices that can be used to make your inference pipelines simpler, more robust, and more efficient."

**Note**: If the main `PCollection` emits inputs and a side input has yet to receive inputs, the main `PCollection` is buffered until there is
            an update to the side input. This could happen with global windowed side inputs with data driven triggers, such as `AfterCount` and `AfterProcessingTime`. Until the side input is updated, emit the default or initial model ID that is used to pass the respective `ModelHandler` as a side input.

> ### Preprocess and postprocess your records before and after inference

Reviewer: I think we could simplify this header. The additional information is in the description.

```suggestion
### Preprocess and postprocess your records
```

> RunInference supports adding pre and post processing operations as part of your transform.

Reviewer: ```suggestion
With RunInference, you can add preprocessing and postprocessing operations to your transform.
```

> To apply preprocessing operations, use `with_preprocess_fn` on your ModelHandler:

Reviewer: ```suggestion
To apply preprocessing operations, use `with_preprocess_fn` on your model handler:
```

```
inference = pcoll | RunInference(model_handler.with_preprocess_fn(lambda x : do_something(x)))
```

> To apply postprocessing operations, use `with_postprocess_fn` on your ModelHandler:

Reviewer: ```suggestion
To apply postprocessing operations, use `with_postprocess_fn` on your model handler:
```

```
inference = pcoll | RunInference(model_handler.with_postprocess_fn(lambda x : do_something_to_result(x)))
```

> You can also chain multiple pre/postprocess operations:

Reviewer: ```suggestion
You can also chain multiple pre- and postprocessing operations:
```

```
inference = pcoll | RunInference(
    model_handler.with_preprocess_fn(
      lambda x : do_something(x)
    ).with_preprocess_fn(
      lambda x : do_something_else(x)
    ).with_postprocess_fn(
      lambda x : do_something_after_inference(x)
    ).with_postprocess_fn(
      lambda x : do_something_else_after_inference(x)
    ))
```

> The preprocessing function will be run before batching/inference and should map your input PCollection

Reviewer: ```suggestion
The preprocessing function is run before batching and inference. This function maps your input `PCollection`
```

> to the base ModelHandler's input type. If you apply multiple preprocessing functions, they will be run on your original

Reviewer: ```suggestion
to the base input type of the model handler. If you apply multiple preprocessing functions, they run on your original
```

> PCollection in order from last applied to first applied.

Reviewer: ```suggestion
`PCollection` in the order of last applied to first applied.
```

> The postprocessing function will be run after inference and should map the base ModelHandler's output

Reviewer: ```suggestion
The postprocessing function runs after inference. This function maps the output type of the base model handler
```

> type to your desired output type. If you apply multiple postprocessing functions, they will be run on your original

Reviewer: ```suggestion
to your desired output type. If you apply multiple postprocessing functions, they run on your original
```

> inference result in order from first applied to last applied.

Reviewer: ```suggestion
inference result in the order of first applied to last applied.
```

### Handle errors while using RunInference

> To handle errors robustly while using RunInference, you can use a _Deadletter Queue (DLQ)_ to output failed records into a separate PCollection for further processing.

Reviewer: ```suggestion
To handle errors robustly while using RunInference, you can use a _dead-letter queue_. The dead-letter queue outputs failed records into a separate `PCollection` for further processing.
```

> This PCollection can then be analyzed and sent to a storage system so that it can be reviewed and resubmitted to the pipeline or discarded.

Reviewer: ```suggestion
This `PCollection` can then be analyzed and sent to a storage system, where it can be reviewed and resubmitted to the pipeline, or discarded.
```

> RunInference has native support for Deadletter Queues; you can use one by applying the `with_exception_handling` to your RunInference transform:

Reviewer: ```suggestion
RunInference has built-in support for dead-letter queues. You can use a dead-letter queue by applying `with_exception_handling` to your RunInference transform:
```

```
main, other = pcoll | RunInference(model_handler).with_exception_handling()
other.failed_inferences | beam.Map(print) # insert logic to handle failed records here
```

> You can also apply this pattern to RunInference transforms with associated pre/postprocessing operations:

Reviewer: ```suggestion
You can also apply this pattern to RunInference transforms with associated pre- and postprocessing operations:
```
